### PR TITLE
Split out curl and winhttp testing on WS2022 to work around CURL bias in test proxy tests

### PR DESCRIPTION
**Pull request by LarryOsterman**
Related Pull Request: #5070 

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 1]
```diff
@@ -202,13 +202,34 @@ std::string GetHeadersAsString(Azure::Core::Http::Request const& request)
...
+
+  for (auto const& header : requestHeaders)
   {
     requestHeaderString += header.first; // string (key)
     requestHeaderString += ": ";
     requestHeaderString += header.second; // string's value
     requestHeaderString += "\r\n";
   }
+
+#if defined(TESTING_BUILD)
```
Comment (antkmsft) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:217]: Why only in tests?
Now we have two different behaviors.

File: eng/pipelines/templates/stages/platform-matrix.json [hunk 2]
```diff
@@ -85,8 +85,13 @@
...
       },
       "TargetPlatform": {
-        "Win32Api_debug_tests": {
-          "CmakeArgs": " -DBUILD_TESTING=ON -DBUILD_PERFORMANCE_TESTS=ON -DRUN_LONG_UNIT_TESTS=ON -DBUILD_TRANSPORT_CURL=ON -DBUILD_TRANSPORT_WINHTTP=ON -DMSVC_USE_STATIC_CRT=ON",
+        "Win32Api_debug_tests_curl": {
+          "CmakeArgs": " -DBUILD_TESTING=ON -DBUILD_PERFORMANCE_TESTS=ON -DRUN_LONG_UNIT_TESTS=ON -DBUILD_TRANSPORT_CURL=ON -DMSVC_USE_STATIC_CRT=ON",
+          "BuildArgs": "--parallel 8 --config Debug",
+          "PublishMapFiles": "true"
+        },
+        "Win32Api_debug_tests_winhttp": {
```
Comment (antkmsft) [inline eng/pipelines/templates/stages/platform-matrix.json:93]: Let's put an explicit `-DBUILD_TRANSPORT_WINHTTP=ON` here?

Comment (LarryOsterman) [reply eng/pipelines/templates/stages/platform-matrix.json:93]: I thought about that, but the default for Windows is to include WINHTTP and I wanted to make sure that was exercised (I don't believe it's used anywhere else).

File: sdk/core/azure-core/inc/azure/core/http/http.hpp [hunk 3]
```diff
@@ -309,7 +309,7 @@ namespace Azure { namespace Core { namespace Http {
      * @brief Get HTTP body as #Azure::Core::IO::BodyStream.
      *
      */
-    Azure::Core::IO::BodyStream* GetBodyStream() { return this->m_bodyStream; }
+    Azure::Core::IO::BodyStream* GetBodyStream() const { return this->m_bodyStream; }
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:312]: Unrelated change - remove from this PR.

It's public surface area and **removing** a `const` would be a breaking change, so we need strong guidance on when we want to use `const` (because we are confident we'll never need to remove it), vs not having it.

Comment (LarryOsterman) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:312]: This *NOT* an unrelated change. It is required to get the code to correctly compile.

Comment (antkmsft) [reply sdk/core/azure-core/inc/azure/core/http/http.hpp:312]: @ahsonkhan Adding const to a method is not a breaking change.
Previously:
```cpp
void f(Azure::Core::Http::Request& req) {
    req.GetBodyStream(); // compiles
}

void g(Azure::Core::Http::Request const& req) {
    req.GetBodyStream(); // wouldn't compile
}
```

Now:
```cpp
void f(Azure::Core::Http::Request& req) {
    req.GetBodyStream(); // compiles
}

void g(Azure::Core::Http::Request const& req) {
    req.GetBodyStream(); // compiles
}
```

I have one suggestion though, I'll drop it in a separate comment.

Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/http/http.hpp:312]:
```suggestion
    Azure::Core::IO::BodyStream* GetBodyStream() { return this->m_bodyStream; }

    /**
     * @brief Get HTTP body as #Azure::Core::IO::BodyStream.
     *
     */
    Azure::Core::IO::BodyStream const* GetBodyStream() const { return this->m_bodyStream; }
```

**Review by antkmsft — APPROVED** [review]
Approving to unblock the CI, and if we are not addressing the same-behavior-as-libcurl in this PR, let's do that in a subsequent one.

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp [hunk 4]
```diff
@@ -202,13 +202,32 @@ std::string GetHeadersAsString(Azure::Core::Http::Request const& request)
...
     requestHeaderString += header.first; // string (key)
     requestHeaderString += ": ";
     requestHeaderString += header.second; // string's value
     requestHeaderString += "\r\n";
   }
+
+  // The test recording infrastructure requires that a Patch verb have a Content-Length header,
+  // because it does not distinguish between requests with and without a body if there's no
+  // Content-Length header.
+  if (request.GetMethod() == HttpMethod::Patch)
```
Comment (Jinming-Hu) [inline sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:220]: Do we want to align with libcurl implementation? 

https://github.com/Azure/azure-sdk-for-cpp/blob/50c18516df110058e1faac27b38360bc56db7c6e/sdk/core/azure-core/src/http/curl/curl.cpp#L428

Libcurl adds `Content-Length: 0` if the request method is not `Get`, `Head`, `Delete`?

**Review by Jinming-Hu — APPROVED** [review]
This looks good to me.

If you don't want to modify production code just to make testing-framework happy, we can also fix this problem with below code (only modifies test-proxy policy).

```Patch
diff --git a/sdk/core/azure-core-test/src/test_proxy_policy.cpp b/sdk/core/azure-core-test/src/test_proxy_policy.cpp
index 766742d4..5e3f3467 100644
--- a/sdk/core/azure-core-test/src/test_proxy_policy.cpp
+++ b/sdk/core/azure-core-test/src/test_proxy_policy.cpp
@@ -85,6 +85,10 @@ std::unique_ptr<RawResponse> TestProxyPolicy::Send(
     // Playback mode
     redirectRequest.SetHeader("x-recording-mode", "playback");
   }
-
+  if (request.GetMethod() == Azure::Core::Http::HttpMethod::Patch
+      && request.GetHeaders().count("Content-Length") == 0)
+  {
+    redirectRequest.SetHeader("Content-Length", "0");
+  }
   return nextHttpPolicy.Send(redirectRequest, ctx);
 }
```

File: sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp (hunk 4 above)
Comment (antkmsft) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:220]: I also think we should align.

Comment (LarryOsterman) [reply sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp:220]: I completely disagree.

Libcurl is a 100% hack which implements the entirety of the HTTP client in our code.
WinHTTP is a production ready HTTP client. We *NEVER* want to second guess the winhttp code.

If there is a disagreement between libcurl and winhttp, WInHTTP wins.
